0025-nat-traversal.adoc

```diff
@@ -0,0 +1,174 @@
+
+- - Requires peers to know peers that are STUN servers
+
+- - Leads to some degree of centralisation
+
+=== TURN
+
+When Bob isn’t able to determine or predict his public facing IP address and port, Bob can dial out to a relay peer, which will keep a long-lived connection open.
+Other peers will be able to dial through the relay peer, which will forward traffic to Bob.
+
+- + Can be used to traverse all commonly know NAT configurations
+
+- + Does not require the Bob to configure his NAT
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires the third peer to know its inbound ports (either through another STUN/TURN or port forwarding)
+
+- - Requires a third peer to serve as a relay for communications between Alice and Bob
+
+- - Requires peers to know peers that are TURN servers
+
+- - Leads to some degree of centralisation
+
+== Recommendation
+
+Assumption: A significant portion of comit nodes will have forwarded ports.
+
+We should implement the following NAT traversal methods in the order listed below.
+After each implementation we should evaluate whether it is necessary to implement the subsequent NAT traversal methods.
+
+1. TURN (essential for maximum NAT coverage)
+2. STUN (optional to reduce load and reliance on TURN servers)
+3. UPnP (optional to reduce load and reliance on TURN servers)
+
+TURN is the first prioirty as it is capable of traversing most commonly known NAT configurations.
+In order for TURN to work, some port forwarded peers will have to operate as TURN servers to allow NAT restricted peers to communicate.
+
+The libp2p specification defines a circuit relay module based off the TURN protocol.
+This protocol allows NAT restricted nodes to "proxy" through other p2p nodes that volunteer to act as p2p-circuit relays.
+These nodes will announce to the network that they are accepting relay connections using libp2p's peer routing.
+Currently there is no Rust implementation of the circuit relay module but the Javascript and Go implementations could be used as a guide.
+
+STUN would be implemented if we found that there are insufficient nodes willing to act as relays due bandwidth concerns.
+STUN allows peers to discover their public IP and ports and establish a communicate directly without a relay.
+
+libp2p has it's own version of the STUN protocol using peer-routing, identify, and reuseport.
+
+1. reuseport uses the same source port for both dialing and listening allowing us to accept new inbound connections on the source port used by the router
+2. identify receives observed address information from all connected peers
+3. peer-routing allows us to publish these addresses to the network and allow peers to discover their own public address on which they can accept inbound connections
+
+If both STUN and TURN are implemented, STUN will be the preferred method and TURN will be used as a fallback if the public address cannot be determined or predicted using STUN.
```

Comment:
Given that COMIT is very much a 1on1 application at the moment, I think the preferred solution should be something where a node can connect to another one out of the blue. What would you propose for that?

Reply:
Without a TURN server or some form of third to relay data we cannot guarantee that COMIT nodes running on mobile networks or the more restrictive NAT configurations will be able to communicate. 

There are publicly available STUN servers would allow for most users to their NAT and connect to each other. [public STUN server list](https://gist.github.com/zziuni/3741933). We could provide a default list like Ethereum does with bootstrap nodes or allow the specify their own list via config or REST API.

[Cases when STUN will fail](https://stackoverflow.com/questions/23655243/nat-traversal-probability-of-success-using-stun)